packages/editor/src/state/callbacks.ts

```diff
@@ -707,20 +714,166 @@ export const useDeleteLocalFile = () =>
         if (!shouldDelete) {
           return;
         }
-        const currentWorkspace = snapshot.getLoadable(
-          currentWorkspaceState,
-        ).contents;
-        // removes from index
-        set(localFilesState, (localFiles) => {
-          const { [id]: removedFile, ...newFiles } = localFiles;
-          return newFiles;
-        });
-        await localforage.removeItem(id);
-        if (currentWorkspace.metadata.id === id) {
-          // set rather than reset to generate new id to avoid id conflicts
-          set(currentWorkspaceState, () => defaultWorkspaceState());
-          reset(diagramState);
+
+        // Delete from cloud storage
+        if (authObject.currentUser != null) {
+          const notif = toast.loading(`Deleting ${name}...`);
+          await deleteDoc(doc(db, authObject.currentUser.uid, id))
+            .catch((error) => {
+              toast.error(`Error deleting diagram: ${name}`);
+              toast.dismiss(notif);
+              return;
+            })
+            .then(() => {
+              // If successful, remove from local state
+              const currentWorkspace = snapshot.getLoadable(
+                currentWorkspaceState,
+              ).contents;
+              set(savedFilesState, (savedFiles) => {
+                const { [id]: removedFile, ...newFiles } = savedFiles;
+                return newFiles;
+              });
+
+              if (currentWorkspace.metadata.id === id) {
+                // set rather than reset to generate new id to avoid id conflicts
+                set(currentWorkspaceState, () => defaultWorkspaceState());
+                reset(diagramState);
+              }
+              toast.dismiss(notif);
+              toast.success(`Deleted ${name}`);
+            });
         }
-        toast.success(`Removed ${name}`);
       },
   );
+
+/**
+ * Used when a "local" workspace is saved for the first time
+ * Also used for duplicate diagram, diagramId passed in as fresh uuid
+ */
+export const useSaveNewWorkspace = () =>
+  useRecoilCallback(({ snapshot, set }) => async (diagramId: string) => {
+    if (
+      authObject.currentUser != null &&
+      authObject.currentUser.uid != undefined
+    ) {
+      const currentWorkspace = snapshot.getLoadable(
+        currentWorkspaceState,
+      ).contents;
+
+      const modificationTime = Date.parse(new Date().toISOString());
+      // Save to cloud
+      const notif = toast.loading("saving...");
+      await setDoc(doc(db, authObject.currentUser.uid, diagramId), {
+        diagramId: diagramId,
+        name: currentWorkspace.metadata.name,
+        lastModified: modificationTime,
+        editorVersion: currentWorkspace.metadata.editorVersion,
+        substance: currentWorkspace.files.substance.contents,
+        style: currentWorkspace.files.style.contents,
+        domain: currentWorkspace.files.domain.contents,
+      })
+        .catch((error) => toast.error("Encountered an error"))
```

Comment:
log the error here if it's meaningful?

Reply:
Error message and error code now reported in console log. 

Error messages not displayed to console because some are preceded by "Firebase: [message]" which felt ugly in the UI.